pkg/amqp/topology_builder.go

```diff
@@ -15,8 +15,9 @@ import (
 // 	config.TopologyBuilder = MyProCustomBuilder{}
 //
 type TopologyBuilder interface {
-	BuildTopology(channel *amqp.Channel, queueName string, exchangeName string, config Config, logger watermill.LoggerAdapter) error
+	BuildTopology(channel *amqp.Channel, queue *amqp.Queue, exchangeName string, config Config, logger watermill.LoggerAdapter) error
 	ExchangeDeclare(channel *amqp.Channel, exchangeName string, config Config) error
+	QueueDeclare(channel *amqp.Channel, queueName string, config Config, logger watermill.LoggerAdapter) (amqp.Queue, error)
```

Comment:
Unfortunately, this is a breaking change. In general, we are trying to avoid such situations and add new features in a way that is backward compatible, so it will not stop working for the current users.

I think that you could add something like the `QueueDeclarer` interface, which may be implemented by `TopologyBuilder` implementations. You can later use [type assertion](https://go.dev/tour/methods/15) on the interface and check if provided topology builder is implementing that interface. But I'm not sure if it will be possible to not touch `BuildTopology` in that way (but it will be required to make it in a backward-compatible way).

An alternative approach may be introducing `TopologyBuilderWithQueueDeclaration` (we can think about a better name :P) and allowing to use one of them to build topology.

The most important part of these approaches is keeping backward compatibility. In theory we can break it, but we would need to release v3 of the library, that I'd like to avoid (we just released v2 :wink:).

Reply:
It would be also good to have a test that checks new kind of builder. It would be enough to have test like that one: https://github.com/ThreeDotsLabs/watermill-amqp/blob/master/pkg/amqp/pubsub_test.go#L188. `tests.TestPublishSubscribe` is the simplest test that should be enough.